### PR TITLE
Recompile test contracts with latest Solidity version + large refactor

### LINKED ISSUE #2301: `enable_strict_bytes_type_checking()` broken with solidty>=0.5.0
### What was wrong?

- `enable_strict_bytes_type_checking()` seems to be broken when compiling with solidity versions `0.5.0` and above. This was discovered when compiling an updated `emitter` contract with updated solidity version in PR #2211.

- Steps to reproduce:  
    - Use solidity `0.5.0` to compile `emitter_old.sol` and the `status` on the transaction receipt to the function will be a `0` (revert). There will be no logs to parse and the test will fail.

### How can it be fixed?

- No details at the moment, but it's looking like `enable_strict_bytes_type_checking()` needs to be updated to be compatible with newer solidity versions.

### PR DESCRIPTION
**Pull request by fselmo**
#### I apologize for the beefiness ahead of time but most of this is refactoring to accommodate the new contract sources.
---

### What was wrong?

closes #2301 
contributes quite a bit to #2460

### How was it fixed?

- Write a script that compiles test contracts in the Solidity version specified and generates test data for testing.
- Move all contracts to the `web3/_utils/contract_sources` folder.
- Compiled contract data is generated in `web3/_utils/contract_sources/contract_data` and neatly assigned to constants for testing.
- Re-compile all contracts using Solidity 0.8.17 (latest at the time of writing).
- Refactor core test fixtures, eliminate fixtures that were not necessary. Use consistent casing for pytest fixture names (this is a large chunk of the refactor).
- Re-generate the integration test fixture to account for bytecode changes in integration tests.
- Add documentation for script and update test docs in general.

I found one improvement along the way due to a failing test:

- When checking a contract's ABI for whether a function is `payable` against a transaction that has a non-zero `value`, we weren't considering the case of `"stateMutability": "nonpayable"`... we were only checking the `payable: {boolean}` part of the ABI. This [change](https://github.com/fselmo/web3.py/blob/recompile-test-contracts-with-latest-solidity/web3/_utils/contracts.py#L372-L373) should clarify what I mean.

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

### CODE REVIEW

**Review by kclowes — APPROVED** [review]
Thank you for this refactor and for the script! Generally, this looks really good. I left a few nitpicks, and I noticed that in some places, we have fixtures for our constants (for ex. `SOME_CONTRACT_BYTECODE`) and sometimes we don't - is there a pattern to leaving those that I didn't pick up on? I commented on it in `tests/core/contracts/test_contract_deployment.py`, but noticed it elsewhere too. Let me know if that doesn't make sense.

File: tests/core/contracts/test_contract_deployment.py [hunk 1]
```diff
@@ -7,10 +7,17 @@
 from web3 import (
     constants,
 )
+from web3._utils.contract_sources.contract_data.constructor_contracts import (
+    CONSTRUCTOR_WITH_ADDRESS_ARGUMENT_CONTRACT_RUNTIME,
```
[the remark this answers is not in the document]
Comment (fselmo) [reply tests/core/contracts/test_contract_deployment.py:11]: The distinction I made was whether or not it's imported in many files as a constant or just one or maybe two test files. I'd be good with making this a session-scoped fixture, or keep it as a constant and reduce some of the clutter in the contracts conftest file. Do you have a preference here? 🤔

[the remark this answers is not in the document]
Comment (kclowes) [reply tests/core/contracts/test_contract_deployment.py:11]: Got it. If it's used in just 1-2 files, I think we can leave it!

File: tests/integration/generate_fixtures/go_ethereum.py [hunk 2]
```diff
@@ -217,8 +219,8 @@ def setup_chain_state(w3):
     # Math Contract
     #
     math_contract_factory = w3.eth.contract(
```
[the remark this answers is not in the document]
Comment (fselmo) [reply tests/integration/generate_fixtures/go_ethereum.py:221]: factory is present in the integration tests, I'm good with it. I can standardize it 👍🏼

**Pull request merged**